sparse/densification.py

```diff
@@ -0,0 +1,140 @@
+import sys
+from numbers import Integral, Number
+from collections import Iterable
+
+import numpy as np
+
+from .utils import _get_broadcast_shape, SparseArray
+
+try:  # Windows compatibility
+    int = long
+except NameError:
+    pass
+
+try:
+    _max_size = sys.maxsize
+except NameError:
+    _max_size = sys.maxint
+
+
+class DensificationConfig(object):
+    def __init__(self, densify=None, max_size=None, min_density=None):
+        if isinstance(densify, (Iterable, DensificationConfig)):
+            self.make_copy_of(densify)
+            return
+
+        if max_size is None:
+            max_size = 10000
+
+        if min_density is None:
+            min_density = 0.25
+
+        if not isinstance(densify, bool) and densify is not None:
+            raise ValueError('always_densify must be a bool or None.')
+
+        if not isinstance(max_size, Integral) or max_size < 0:
+            raise ValueError("max_nnz must be a non-negative integer.")
+
+        if not isinstance(min_density, Number) or not (0.0 <= min_density <= 1.0):
+            raise ValueError('min_density must be a number between 0 and 1.')
+
+        self.densify = densify
+        self.max_size = int(max_size)
+        self.min_density = float(min_density)
+
+    def make_copy_of(self, other):
+        if isinstance(other, Iterable):
+            single = DensificationConfig.from_many(other)
+        else:
+            single = other
+
+        self.densify = single.densify
+        self.max_size = single.max_size
+        self.min_density = single.min_density
+
+    def _should_densify(self, size, density):
+        if self.densify:
+            return True
+        elif self.densify is False:
+            return False
+        else:
+            return self.max_size >= size or self.min_density <= density
+
+    def _raise_if_fails(self, size, density, name):
+        if not self._should_densify(size, density):
+            raise ValueError("Performing this operation would produce "
+                             "a dense result: %s" % name)
+
+    def check(self, name, *arrays):
+        result_shape = ()
+        density = 1.0
+
+        for arr in arrays:
+            if isinstance(arr, SparseArray):
+                density = min(density, arr.density)
+                result_shape = _get_broadcast_shape(result_shape, arr.shape)
+
+        size = np.prod(result_shape, dtype=np.uint64)
+
+        self._raise_if_fails(size, density, name)
+
+    @staticmethod
+    def validate(configs):
+        if isinstance(configs, Iterable):
+            for config in configs:
+                DensificationConfig._validate_single(config)
+
+            return
+
+        DensificationConfig._validate_single(configs)
+
+    @staticmethod
+    def _validate_single(config):
+        if not isinstance(config, DensificationConfig):
+            raise ValueError('Invalid DensificationConfig.')
+
+    @staticmethod
+    def combine(*configs):
```

Comment:
What's the difference between `combine` and `from_many`?

Reply:
I `combine`d them (pun intended).

Seriously, though. `from_parents` builds the parents, and reduces if none of the parents are in a context manager mode.

`_reduce_from_parents` gets the current rules without regards to what the parents are.